README.md
<div align='center'>

![img](img/Group%204.png)

![npm](https://badgen.net/npm/v/math-pro)
![install size](https://badgen.net/packagephobia/install/math-pro)
![status](https://badgen.net/github/status/pro-js/math-pro)
![stars](https://badgen.net/github/stars/pro-js/math-pro)
![downloads](https://badgen.net/npm/dt/math-pro)
![last commit](https://badgen.net/github/last-commit/pro-js/math-pro)

</div>


## Introduction

Math Pro is an extensive math library which will help you to do the basic calculation of math. Besides basic calculation you can also do the sort like sort, search. Some number properties of math have also included here. If you fear math then this library is for you.

## Features

- Support numbers, big numbers, arrays
- Is compatible with JavaScript's built-in functions.
- Open source

## Download and usage

To use this package you need to install it from [npm](https://www.npmjs.com/package/math-pro). In your project directory write the following command to get the package -

`npm i math-pro`


After installation you need to import the package in your `.js` file.

#### import as ES5 format

After installation you need to import the package to use.


```js
const mt = require("math-pro");

console.log(mt.addition(2, 4, 5, 6));

// 17
```

#### import as ES6 format 

```js
import mt from 'math-pro'

console.log(mt.addition(2, 4, 5, 6));

// 17
```
_**N.B**: You may encounter an issue when importing as ES6 format. To fix the issue in your **package.json** file add `"type": "module"`_
## Available functions

Below are the available functions that we have in our library.

`:x: means you can't pass variable in that format`

### Basic functions

|    Function name     |     Example (Arguments)     |        Example (Array)        |
| :------------------: | :-------------------------: | :---------------------------: |
|    **addition()**    |    addition(2,3,4) // 9     |    addition([2,3,4]) // 9     |
|  **subtraction()**   |  subtraction(2,3,4) // -5   |  subtraction([2,34,4]) // -5  |
| **multiplication()** | multiplication(2,3,4) // 24 | multiplication([2,3,4]) // 24 |
|    **division()**    |  division(8,2,3) // 1.3333  |  division([2,3,4]) // 1.3333  |
|    **average()**     |     average(2,3,4) // 3     |     average([2,3,4]) // 3     |
|  **getRemainder()**  |   getRemainder(5, 2) // 1   |              :x:              |
|   **getMinimum()**   |   getMinimum(2,3,4) // 2    |   getMinimum([2,3,4]) // 2    |
|   **getMaximum()**   |   getMaximum(2,3,4) // 4    |   getMaximum([2,3,4]) // 4    |

### Conversion functions

|    Function name     |         Example (Arguments)          | Example (Array) |
| :------------------: | :----------------------------------: | :-------------: |
|  **digitToArray()**  |     digitToArray(234) // [1,2,3]     |       :x:       |
| **numberToChars()**  |       numberToChars(51) // AZ        |       :x:       |
|   **degsToRads()**   | degsToRads(5) // 0.08726646259971647 |       :x:       |
|   **radsToDegs()**   | radsToDegs(0.08726646259971647) // 5 |       :x:       |
| **stringToNumber()** |      stringToNumber("25") // 25      |       :x:       |
| **convertNumber()**  | convertNumber('100111', 2, 16) // 27 |       :x:       |

### Check functions

|   Function name    |    Example (Arguments)    | Example (Array) |
| :----------------: | :-----------------------: | :-------------: |
|  **isDivisor()**   | isDivisor(25, 5) // true  |       :x:       |
|    **isEven()**    |    isEven(10) // true     |       :x:       |
|    **isOdd()**     |    isOdd(10) // false     |       :x:       |
|   **isPrime()**    |    isPrime(7) // true     |       :x:       |
| **isPalindrome()** | isPalindrome(232) // true |       :x:       |
|  **isPositive()**  |  isPositive(50) // true   |       :x:       |
|  **isNegative()**  |  isNegative(-50) // true  |       :x:       |
| **isPowerOfTwo()** | isPowerOfTwo(32) // true  |       :x:       |

### Number functions

|    Function name     |         Example (Arguments)         |  Example (Array)   |
| :------------------: | :---------------------------------: | :----------------: |
|      **gcd()**       |          gcd(16,4,8) // 4           | gcd([16,4,8]) // 4 |
|     **round()**      |     round(1.23456, 3) // 1.235      |        :x:         |
|    **toFixed()**     |    toFixed(1.23456, 3) // 1.234     |        :x:         |
|  **getFactorial()**  |       getFactorial(5) // 120        |        :x:         |
|    **bigPower()**    |         bigPower(3,3) // 27         |        :x:         |
|  **fibSequence()**   |    fibSequence(3) // [ 1, 1, 2 ]    |        :x:         |
|     **fibNth()**     |          fibNth(10) // 55           |        :x:         |
| **pascalTriangle()** | pascalTriangle(3) // [ 1, 3, 3, 1 ] |        :x:         |
|  **getPrimeList()**  |   getPrimeList(5) // [ 2, 3, 5 ]    |        :x:         |
|  **randomNumber()**  |      randomNumber(3, 10) // 9       |        :x:         |
| **reverseNumber()**  |      reverseNumber(234) // 432      |        :x:         |

### Search functions

|   Function name    | Example (Arguments) |        Example (Array)        |
| :----------------: | :-----------------: | :---------------------------: |
| **binarySearch()** |         :x:         | binarySearch([2,3,4], 4) // 2 |
| **linearSearch()** |         :x:         | linearSearch([3,4,2], 4) // 1 |

### Sort functions

|    Function name    | Example (Arguments) |             Example (Array)              |
| :-----------------: | :-----------------: | :--------------------------------------: |
|  **bubbleSort()**   |         :x:         |      bubbleSort([3,2,4]) // [2,3,4]      |
|   **mergeSort()**   |         :x:         |   mergeSort([3,2,4], "des") // [4,3,2]   |
|   **quickSort()**   |         :x:         |      quickSort([3,2,4]) // [2,3,4]       |
| **selectionSort()** |         :x:         | selectionSort([3,2,4], "asc") // [2,3,4] |
| **insertionSort()** |         :x:         | insertionSort([3,2,4], "des") // [4,3,2] |

#### This project is licensed under `MIT`


HISTORY.md
# History

## 2020-07-14, version 1.0.0

- First publish of the math-pro package.


lib/number.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Pass an (array or argument)
 * return greatest common divisor
 * @param {*} array
 * gcd(4, 7); // 1
 */
var gcd = function () {
    var array = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        array[_i] = arguments[_i];
    }
    if (array[0][0] !== undefined)
        array = array[0];
    var ans = array[0];
    for (var i = 1; i < array.length; i++) {
        ans = gcdCalculate(ans, array[i]);
    }
    return ans;
};
var gcdCalculate = function (firstNumber, secondNumber) {
    return secondNumber === 0 ? firstNumber
        : gcdCalculate(secondNumber, firstNumber % secondNumber);
};
/**
 * Round a number to a given number of digits
 * @param {*} number
 * @param {*} decimals
 * round(1.234567, 3); // 1.235
 * round(1.2345678, 5); // 1.23457

const round = (number:any, decimals = 0) => {
  return Number(`${Math.round(`${number}e${decimals}`)}e-${decimals}`);
};


 * Truncate a number to a given number of decimal places without rounding
 * @param {*} number
 * @param {*} fixed
 * toFixed(25.198726354, 2); // 25.19
 * toFixed(25.198726354, 5); // 25.19872

const toFixed = (number:any, fixed:any):any => {
  return `${number}`.match(new RegExp(`^-?\\d+(?:\.\\d{0,${fixed}})?`))[0];
};
 */
/**
 * Pass a number and calculate the factorial of the number
 * @param {*} number
 */
var getFactorial = function (number) {
    return number != 1 ? number * getFactorial(number - 1) : 1;
};
/**
 * Pass a base number, power number and calculate the power of the number
 * @param {*} number
 */
var bigPower = function (base, power) {
    if (power === 0)
        return 1;
    if (power % 2 === 0) {
        var multiplier_1 = bigPower(base, power / 2);
        return multiplier_1 * multiplier_1;
    }
    var multiplier = bigPower(base, Math.floor(power / 2));
    return multiplier * multiplier * base;
};
/**
 * Return a fibonacci sequence as an array.
 * @param number
 * @return {number[]}
 * console.log(fibSequence(10)) // [ 1, 1, 2, 3, 5, 8, 13, 21, 34, 55 ]
 */
var fibSequence = function (number) {
    var fibSeq = [1], currentValue = 1, previousValue = 0;
    if (number === 1)
        return fibSeq;
    var iterationsCounter = number - 1;
    while (iterationsCounter) {
        currentValue += previousValue;
        previousValue = currentValue - previousValue;
        fibSeq.push(currentValue);
        iterationsCounter -= 1;
    }
    return fibSeq;
};
/**
 * Calculate fibonacci number at specific position using Dynamic Programming approach.
 * @param number
 * @return {number}
 * console.log(fibonacciNth(10)) // 55
 */
var fibNth = function (number) {
    var currentValue = 1, previousValue = 0;
    if (number === 1)
        return 1;
    var iterationsCounter = number - 1;
    while (iterationsCounter) {
        currentValue += previousValue;
        previousValue = currentValue - previousValue;
        iterationsCounter -= 1;
    }
    return currentValue;
};
/**
 * Get pascal Triangle
 * @param {number} lineNumber - zero based.
 * @return {number[]}
 * console.log(pascalTriangle(10))
 * [ 1, 10, 45, 120, 210, 252, 210, 120, 45, 10, 1 ]
 */
var pascalTriangle = function (lineNumber) {
    var currentLine = [1], currentLineSize = lineNumber + 1;
    for (var numIndex = 1; numIndex < currentLineSize; numIndex += 1) {
        currentLine[numIndex] =
            (currentLine[numIndex - 1] * (lineNumber - numIndex + 1)) / numIndex;
    }
    return currentLine;
};
/**
 * Get prime list
 * @param {number}
 * @return {primes[]}
 * console.log(getPrimeList(20)) // [ 2, 3, 5, 7, 11, 13, 17, 19 ]
 */
var getPrimeList = function (number) {
    var isPrime = new Array(number + 1), primes = [];
    isPrime[0] = true;
    isPrime[1] = true;
    for (var num = 2; num <= number; num += 1) {
        if (isPrime[num] === undefined) {
            primes.push(num);
            var nextnum = num * num;
            while (nextnum <= number) {
                isPrime[nextnum] = true;
                nextnum += num;
            }
        }
    }
    return primes;
};
/**
 * Pass minimum and maximum number to generate the random number between them
 * randomNumber(3, 10) // 10
 * randomNumber(3, 10) // 8
 * @param {*} min
 * @param {*} max
 */
var randomNumber = function (min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
};
/**
 * Pass a number and get the reverse of that number with actual sign
 * reverseNumber(32) // 23
 * reverseNumber(-32) // -23
 * @param {*} number
 */
var reverseNumber = function (number) {
    var sign = false;
    if (number < 0) {
        sign = true;
        number *= -1;
    }
    var reversed = number.toString().split("").reverse().join("");
    if (sign)
        reversed *= -1;
    return reversed;
};
exports.default = {
    gcd: gcd,
    getFactorial: getFactorial,
    bigPower: bigPower,
    fibSequence: fibSequence,
    fibNth: fibNth,
    pascalTriangle: pascalTriangle,
    getPrimeList: getPrimeList,
    randomNumber: randomNumber,
    reverseNumber: reverseNumber,
};
